### PR TITLE
Campaign Join Modal and Form

### LINKED ISSUE #387: Campaign join form
## Description

See #251 for narrative.
There should be a form that allows users to join a campaign. It should have a field where the user can enter a campaign join code. This field can be prefilled if the user uses a join link. The page must show the characters a player can join as. It should also present the option to join as a GM instead. If a user chooses to join as a GM then they should not be able to choose a character

### PR DESCRIPTION
**Pull request by AnnieTran13**
closes #387 

- Join button in campaign index should open a modal asking for user's campaign invite code
- After joining with invite code, the user should be redirected to the join form where they are asked to assign a character or check if they are the GM

### CODE REVIEW

File: resources/js/Components/CampaignJoinModal/CampaignJoinModal.tsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    message: string
+}
+
+type CampaignJoinModalDataType = {
+    code: string | ''
+}
+
+const CampaignJoinModal = ({open, onClose, message}: CampaignJoinModalPropType) => {
+    const CAMPAIGN_JOIN_MODAL_FORM_INITIAL_VALUE: CampaignJoinModalDataType = {
+        code: '',
```
Comment (willyyhuang) [inline resources/js/Components/CampaignJoinModal/CampaignJoinModal.tsx:39]: use `undefined` instead

File: resources/js/Components/CampaignJoinModal/CampaignJoinModal.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+`
+
+type CampaignJoinModalPropType = {
+    open: boolean
+    onClose: () => void
+    message: string
+}
+
+type CampaignJoinModalDataType = {
+    code: string | ''
```
Comment (willyyhuang) [inline resources/js/Components/CampaignJoinModal/CampaignJoinModal.tsx:34]: same here

File: resources/js/Components/CampaignJoinModal/CampaignJoinModal.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            BackdropComponent={Backdrop}
+            BackdropProps={{
+                timeout: 500,
+            }}
+            aria-label='campaign-join-modal'>
+            <Fade in={open}>
+                <StyledBox borderRadius={2}>
+                    <Grid container rowSpacing={2} alignItems='center'>
+                        <Grid item xs={12}>
+                            <Typography sx={{color: 'white'}}>{message}</Typography>
```
Comment (willyyhuang) [inline resources/js/Components/CampaignJoinModal/CampaignJoinModal.tsx:59]: use `color` in theme provider, stick to the design color theme, I believe white is `secondary`

File: resources/js/Components/Form/CampaignRegistrationForm/CampaignRegistrationForm.tsx [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    characters,
+}: CampaignRegistrationFormPropType) => {
+    const CAMPAIGN_REGISTRATION_FORM_INITIAL_VALUE: CampaignRegistrationFormDataType = {
+        character_id: null,
+    }
+
+    const {data, setData, errors} = useForm<CampaignRegistrationFormDataType>(
+        CAMPAIGN_REGISTRATION_FORM_INITIAL_VALUE,
+    )
+    console.log(campaign)
```
Comment (willyyhuang) [inline resources/js/Components/Form/CampaignRegistrationForm/CampaignRegistrationForm.tsx:28]: remove console.log

File: resources/js/Components/Form/CampaignRegistrationForm/CampaignRegistrationForm.tsx [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+type CampaignRegistrationFormDataType = {
+    character_id: number | null
+}
+
+const CampaignRegistrationForm = ({
+    campaign,
+    characters,
+}: CampaignRegistrationFormPropType) => {
+    const CAMPAIGN_REGISTRATION_FORM_INITIAL_VALUE: CampaignRegistrationFormDataType = {
```
Comment (willyyhuang) [inline resources/js/Components/Form/CampaignRegistrationForm/CampaignRegistrationForm.tsx:21]: no point of declaring a constant if it's only going to be used once or no conditional data

File: resources/js/Components/Form/CampaignRegistrationForm/CampaignRegistrationForm.tsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+    const {data, setData, errors} = useForm<CampaignRegistrationFormDataType>(
+        CAMPAIGN_REGISTRATION_FORM_INITIAL_VALUE,
+    )
+    console.log(campaign)
+    return (
+        <Grid container columnSpacing={5} rowSpacing={5}>
+            <Grid item xs={12}>
+                <Typography>
+                    You are joining the &quot;
```
Comment (willyyhuang) [inline resources/js/Components/Form/CampaignRegistrationForm/CampaignRegistrationForm.tsx:33]: just use string notation with variable

File: resources/js/Components/Form/CampaignRegistrationForm/CampaignRegistrationForm.tsx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                        value={data.character_id}
+                        onChange={(e) =>
+                            setData('character_id', parseInt(e.target.value))
+                        }
+                        options={characters}
+                    />
+                    {errors?.character_id && <ErrorText message={errors?.character_id} />}
+                </Grid>
+                <Grid item xs={6}>
+                    <FormControlLabel
```
Comment (willyyhuang) [inline resources/js/Components/Form/CampaignRegistrationForm/CampaignRegistrationForm.tsx:53]: this is not connected to the form

File: resources/js/Components/CampaignJoinModal/CampaignJoinModal.tsx (hunk 3 above)
Comment (AnnieTran13) [reply resources/js/Components/CampaignJoinModal/CampaignJoinModal.tsx:59]: secondary is a very light green but I think it looks good with it so i'll keep it that way.

**Pull request closed without merging**